### PR TITLE
Start collecting per-site stats of client versions.

### PR DESCRIPTION
**Pull request by lionell-pack-ttd**
image

### CODE REVIEW

File: src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java [hunk 1]
```diff
@@ -3304,7 +3305,7 @@ void cstgDomainNameCheckFailsAndLogInvalidHttpOrigin(String httpOrigin, Vertx ve
                     assertFalse(respJson.containsKey("body"));
                     assertEquals("unexpected http origin", respJson.getString("message"));
                     assertEquals("invalid_http_origin", respJson.getString("status"));
-                    Assertions.assertTrue(logWatcher.list.get(0).getFormattedMessage().contains("InvalidHttpOriginAndAppName: site test (123): http://gototest.com"));
+                    assertThat(logWatcher.list.stream().map(ILoggingEvent::getFormattedMessage).collect(Collectors.toList())).contains("InvalidHttpOriginAndAppName: site test (123): http://gototest.com");
```
Comment (lionell-pack-ttd) [inline src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:3308]: Is this test flaky? It failed once, but I haven't seen it fail again. Hopefully this change will give us more info if it fails again.

File: src/main/java/com/uid2/operator/model/StatsCollectorMessageItem.java [hunk 2]
```diff
@@ -48,4 +50,7 @@ public Integer getSiteId() {
     public void setSiteId(Integer siteId) {
         this.siteId = siteId;
     }
+
```
[the remark this answers is not in the document]
Comment (lionell-pack-ttd) [reply src/main/java/com/uid2/operator/model/StatsCollectorMessageItem.java:53]: I'll run a reformat on the file and I'll check it in as IntelliJ does it.

File: src/main/java/com/uid2/operator/monitoring/ClientVersionStatRecorder.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        this.siteClientBucketLimit = maxVersionBucketsPerSite;
+    }
+
+    public Stream<SiteClientVersionStat> getStatsView() {
+        return siteIdToVersionCounts.entrySet().stream().map(entry -> new SiteClientVersionStat(entry.getKey(), entry.getValue()));
+    }
+
+    private void removeLowVersionCounts(int siteId) {
+        var versionCounts = siteIdToVersionCounts.get(siteId);
+        if (versionCounts == null) return;
```
[the remark this answers is not in the document]
Comment (lionell-pack-ttd) [reply src/main/java/com/uid2/operator/monitoring/ClientVersionStatRecorder.java]: Fixed.

File: src/main/java/com/uid2/operator/monitoring/ClientVersionStatRecorder.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                .toList();
+        for (var entry : lowestEntries) {
+            var notRecordedCount = versionCounts.getOrDefault(NOT_RECORDED, 0);
+            versionCounts.put(NOT_RECORDED, notRecordedCount + entry.getValue());
+            versionCounts.remove(entry.getKey());
+        }
+    }
+
+    public void add(Integer siteId, String clientVersion) {
+        if (siteId == null || clientVersion == null) {
```
[the remark this answers is not in the document]
Comment (lionell-pack-ttd) [reply src/main/java/com/uid2/operator/monitoring/ClientVersionStatRecorder.java]: Good point - probably no point logging the number of empty strings we see. Fixed.

File: src/main/java/com/uid2/operator/monitoring/StatsCollectorVerticle.java [hunk 5]
```diff
@@ -19,11 +19,14 @@
 import java.time.Instant;
 import java.util.*;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
 
 public class StatsCollectorVerticle extends AbstractVerticle implements IStatsCollectorQueue {
     private static final Logger LOGGER = LoggerFactory.getLogger(StatsCollectorVerticle.class);
     private HashMap<String, EndpointStat> pathMap;
 
+    private ClientVersionStatRecorder clientVersionStat;
```
[the remark this answers is not in the document]
Comment (lionell-pack-ttd) [reply src/main/java/com/uid2/operator/monitoring/StatsCollectorVerticle.java]: Fixed.

File: src/main/java/com/uid2/operator/monitoring/StatsCollectorVerticle.java [hunk 6]
```diff
@@ -148,12 +153,12 @@ public void handleMessage(Message message) {
...
 
-    private Void serializeToLogs(Object[] stats) {
+    private Void serializeToLogs(List<Object> stats) {
         LOGGER.debug("Starting JSON Serialize");
-        ObjectMapper mapper = new ObjectMapper();
+        ObjectMapper statMapper = new ObjectMapper();
         for (Object stat : stats) {
             try {
-                String jsonString = mapper.writeValueAsString(stat);
+                String jsonString = statMapper.writeValueAsString(stat);
```
[the remark this answers is not in the document]
Comment (lionell-pack-ttd) [reply src/main/java/com/uid2/operator/monitoring/StatsCollectorVerticle.java]: I've updated the new format - here's the log lines the tests look for:
```
version log; siteId=1 versions={"uid2-sdk-3.0.01":1,"uid2-sdk-3.0.02":1,"uid2-sdk-3.0.00":1}
version log; siteId=1 versions={"uid2-sdk-3.0.01":2,"uid2-sdk-3.0.02":1,"uid2-sdk-3.0.00":1}
version log; siteId=2 versions={"<Not recorded>":21,"uid2-sdk-3.0.06":7,"uid2-sdk-3.0.07":8,"uid2-sdk-3.0.011":12,"uid2-sdk-3.0.08":9,"uid2-sdk-3.0.010":11,"uid2-sdk-3.0.09":10,"uid2-sdk-3.0.0single":1
```

Is that right? Will the <> in the "<Not recorded>" entry cause a problem for your parsing?

[the remark this answers is not in the document]
Comment (Ian-Nara) [reply src/main/java/com/uid2/operator/monitoring/StatsCollectorVerticle.java]: It will work, tested your branch locally

**Review by thomasm-ttd — APPROVED** [review]

**Pull request merged**